core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,241 @@
+    ]
+
+    from . import seed
+
+AUXILIARY_DATA_HASH_SIZE = 32
+CATALYST_VOTING_PUBLIC_KEY_LENGTH = 32
+CATALYST_REGISTRATION_HASH_SIZE = 32
+
+METADATA_KEY_CATALYST_REGISTRATION = 61284
+METADATA_KEY_CATALYST_REGISTRATION_SIGNATURE = 61285
+
+
+def validate_auxiliary_data(
+    keychain: seed.Keychain,
+    auxiliary_data: CardanoTxAuxiliaryDataType | None,
+    protocol_magic: int,
+    network_id: int,
+) -> None:
+    if not auxiliary_data:
+        return
+
+    if auxiliary_data.blob:
+        _validate_auxiliary_data_blob(auxiliary_data.blob)
+    elif auxiliary_data.catalyst_registration_parameters:
+        _validate_catalyst_registration_parameters(
+            keychain,
+            auxiliary_data.catalyst_registration_parameters,
+            protocol_magic,
+            network_id,
+        )
+    else:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_auxiliary_data_blob(auxiliary_data_blob: bytes) -> None:
+    try:
+        # validation to prevent CBOR injection and invalid CBOR
+        # we don't validate data format, just that it's a valid CBOR
+        cbor.decode(auxiliary_data_blob)
+    except Exception:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_catalyst_registration_parameters(
+    keychain: seed.Keychain,
+    catalyst_registration_parameters: CardanoCatalystRegistrationParametersType | None,
+    protocol_magic: int,
+    network_id: int,
+) -> None:
+    if not catalyst_registration_parameters:
+        raise INVALID_AUXILIARY_DATA
+
+    if (
+        len(catalyst_registration_parameters.voting_public_key)
+        != CATALYST_VOTING_PUBLIC_KEY_LENGTH
+    ):
+        raise INVALID_AUXILIARY_DATA
+
+    if not SCHEMA_STAKING_ANY_ACCOUNT.match(
+        catalyst_registration_parameters.staking_path
+    ):
+        raise INVALID_AUXILIARY_DATA
+
+    address_parameters = catalyst_registration_parameters.reward_address_parameters
+    if address_parameters.address_type == CardanoAddressType.BYRON:
+        raise INVALID_AUXILIARY_DATA
+
+    # try to derive the address to validate it
+    derive_address_bytes(keychain, address_parameters, protocol_magic, network_id)
+
+
+async def show_auxiliary_data(
+    ctx: wire.Context,
+    keychain: seed.Keychain,
+    auxiliary_data: CardanoTxAuxiliaryDataType | None,
+    protocol_magic: int,
+    network_id: int,
+) -> None:
+    if not auxiliary_data:
+        return
+
+    if auxiliary_data.catalyst_registration_parameters:
+        await _show_catalyst_registration(
+            ctx,
+            keychain,
+            auxiliary_data.catalyst_registration_parameters,
+            protocol_magic,
+            network_id,
+        )
+
+    auxiliary_data_bytes = get_auxiliary_data_cbor(
+        keychain, auxiliary_data, protocol_magic, network_id
+    )
+
+    auxiliary_data_hash = hash_auxiliary_data(bytes(auxiliary_data_bytes))
+    await show_auxiliary_data_hash(ctx, auxiliary_data_hash)
```

Comment:
wondering if this is necessary if you show the parsed parameters. but feel free to keep it if you want

Reply:
I think it's better to show it to the users, just in case they might want to verify the hash.